grammar.js

```diff
@@ -1080,22 +1086,30 @@ module.exports = grammar({
       ),
       seq(
         $.keyword_as,
-        alias($._literal_string, $.literal),
+        alias(
+          choice(
+            $._single_quote_string,
+            $._double_quote_string,
+          ),
+          $.literal
+        ),
       ),
       seq(
         $.keyword_as,
         $.dollar_quote,
         $._function_body_statement,
+        optional(';'),
         $.dollar_quote,
       ),
     ),
 
     function_language: $ => seq(
       $.keyword_language,
-      choice(
-        $.keyword_sql,
-        $.keyword_plpgsql,
-      ),
+      // TODO Maybe we should do different version of function_body_statement in
+      // regard to the defined language to match either sql, plsql or
+      // plpgsql. Currently the function_body_statement support only sql.  And
+      // maybe for other language the function_body should be a string.
```

Comment:
sql functions without `begin atomic` are strings already, but it's useful to parse them anyway. Other sql-derived dialects are also useful to parse (I'm leaning on that with pl/pgsql in [pdot](https://gitlab.com/dmfay/pdot)) but full support is a question mark atm, see also https://github.com/DerekStride/tree-sitter-sql/pull/141#discussion_r1187812386

Reply:
Currently the inside of `$.dollar_quote` ... `$.dollar_quote` will not parse the first time you put a (pg)plsql for loop for instance.
One easy solution is to parse the whole body as a string, but the highlight will be one of a string.

```sql
CREATE FUNCTION public.update_time() RETURNS trigger
  LANGUAGE plpgsql
AS $$
BEGIN
  NEW.updated_at = CURRENT_TIMESTAMP;
  RETURN NEW;
END;
$$;
```

```
(program
 (statement
  alias: 
   (create_function (keyword_create) (keyword_function)
    (object_reference schema: (identifier) . name: (identifier))
    ( ) (keyword_returns) (keyword_trigger)
    (function_language (keyword_language) (identifier))
    (function_body (keyword_as) (dollar_quote) (keyword_begin)
     (ERROR . = (keyword_current_timestamp) ;)
     (keyword_return)
     (field name: (identifier))
     ; (keyword_end) ; (dollar_quote))))
 ;)
```

Reply:
An other solution might be:

```
    function_body: $ => choice(
      seq(
        $._function_return,
        ';'
      ),
      seq(
        $.keyword_begin,
        $.keyword_atomic,
        repeat1(
          seq(
            $._function_body_statement,
            ';',
          ),
        ),
        $.keyword_end,
      ),
      seq(
        $.keyword_as,
        $.dollar_quote,
        optional(
          seq(
            $.keyword_declare,
            repeat1(
              $.function_declaration,
            ),
          ),
        ),
        $.keyword_begin,
        repeat($.junk)
        $.keyword_end,
        optional(';'),
        $.dollar_quote,
      ),
      seq(
        $.keyword_as,
        alias(
          choice(
            $._single_quote_string,
            $._double_quote_string,
          ),
          $.literal
        ),
      ),
    ),

    junk: $ => choice(
      $.all_keyword_expect_end,
      $.object_reference,
      $.all_operator,
      $.all_punctuation,
      $.literal,
      wrapped_in_parenthesis($.junk),
      wrapped_in_bracket($.junk),
      wrapped_in_curly_bracket($.junk),
    ),
```
It match all without structure but give you base highlight capability (like keyword, number, operator...).

Reply:
I actually need the "best effort" parsing of pl/pgsql -- even with ERROR nodes it still recognizes regular queries and DML statements. Try it with longer function bodies. Here's a function with `if x then (update statement); end if; functioncall()`:

```scheme
(program [0, 0] - [11, 0]
  (statement [0, 0] - [10, 10]
    (create_function [0, 0] - [10, 10]
      (keyword_create [0, 0] - [0, 6])
      (keyword_or [0, 7] - [0, 9])
      (keyword_replace [0, 10] - [0, 17])
      (keyword_function [0, 18] - [0, 26])
      (object_reference [0, 27] - [0, 47]
        schema: (identifier [0, 27] - [0, 39])
        name: (identifier [0, 40] - [0, 47]))
      (function_arguments [0, 47] - [0, 49])
      (keyword_returns [1, 1] - [1, 8])
      (keyword_trigger [1, 9] - [1, 16])
      (function_language [2, 1] - [2, 17]
        (keyword_language [2, 1] - [2, 9])
        (keyword_plpgsql [2, 10] - [2, 17]))
      (function_body [3, 0] - [10, 10]
        (keyword_as [3, 0] - [3, 2])
        (dollar_quote [3, 3] - [3, 13])
        (keyword_begin [3, 14] - [3, 19])
        (ERROR [4, 2] - [4, 15]
          (keyword_then [4, 11] - [4, 15]))
        (statement [5, 4] - [5, 65]
          (update [5, 4] - [5, 65]
            (keyword_update [5, 4] - [5, 10])
            (relation [5, 11] - [5, 32]
              (object_reference [5, 11] - [5, 32]
                schema: (identifier [5, 11] - [5, 23])
                name: (identifier [5, 24] - [5, 32])))
            (keyword_set [5, 33] - [5, 36])
            (assignment [5, 37] - [5, 65]
              left: (field [5, 37] - [5, 41]
                name: (identifier [5, 37] - [5, 41]))
              right: (literal [5, 44] - [5, 65]))))
        (keyword_end [6, 2] - [6, 5])
        (ERROR [6, 6] - [9, 3])
        (dollar_quote [10, 0] - [10, 10])))))
```

the errors don't stop it from processing the rest and returning useful output, although in this case the select statement is not recognized after the `end if`.